router/types/types.go

```diff
@@ -1,88 +1,138 @@
+	// Type is the type of Route, either "http" or "tcp".
+	Type string `json:"type"`
+	// ID is the unique ID of this route.
+	ID string `json:"id,omitempty"`
+	// ParentRef is ... TODO
 	ParentRef string `json:"parent_ref,omitempty"`
-	Type      string `json:"type,omitempty"`
+	// Service is the ID of the service.
+	Service string `json:"service"`
+	// CreatedAt is the time this Route was created.
+	CreatedAt time.Time `json:"created_at,omitempty"`
+	// UpdatedAt is the time this Route was last updated.
+	UpdatedAt time.Time `json:"updated_at,omitempty"`
 
-	Config *json.RawMessage `json:"config,omitempty"`
+	// Domain is the domain name of this Route. It is only used for HTTP routes.
+	Domain string `json:"domain,omitempty"`
+	// TLSCert is the optional TLS public certificate of this Route. It is only
+	// used for HTTP routes.
+	TLSCert string `json:"tls_cert,omitempty"`
+	// TLSCert is the optional TLS private key of this Route. It is only
+	// used for HTTP routes.
+	TLSKey string `json:"tls_key,omitempty"`
+	// Sticky is whether or not to use sticky sessions for this route. It is only
+	// used for HTTP routes.
+	Sticky bool `json:"sticky,omitempty"`
 
-	CreatedAt *time.Time `json:"created_at,omitempty"`
-	UpdatedAt *time.Time `json:"updated_at,omitempty"`
+	// Port is the TCP port to listen on for TCP Routes.
+	Port int32 `json:"port,omitempty"`
 }
 
-var ErrWrongType = errors.New("router: the requested route type does not match the actual type")
-var ErrNoConfig = errors.New("router: the supplied route has no configuration")
+func (r Route) HTTPRoute() *HTTPRoute {
+	return &HTTPRoute{
+		ID:        r.ID,
+		ParentRef: r.ParentRef,
+		Service:   r.Service,
+		CreatedAt: r.CreatedAt,
+		UpdatedAt: r.UpdatedAt,
 
-func (r *Route) HTTPRoute() *HTTPRoute {
-	rCopy := *r
-	if rCopy.Config == nil {
-		empty := json.RawMessage(`{}`)
-		rCopy.Config = &empty
+		Domain:  r.Domain,
+		TLSCert: r.TLSCert,
+		TLSKey:  r.TLSKey,
+		Sticky:  r.Sticky,
 	}
-	route := &HTTPRoute{Route: &rCopy}
-	route.Route.Config = nil
-	json.Unmarshal(*r.Config, route)
-	return route
 }
 
-func (r *Route) TCPRoute() *TCPRoute {
-	rCopy := *r
-	if rCopy.Config == nil {
-		empty := json.RawMessage(`{}`)
-		rCopy.Config = &empty
+func (r Route) TCPRoute() *TCPRoute {
+	return &TCPRoute{
+		ID:        r.ID,
+		ParentRef: r.ParentRef,
+		Service:   r.Service,
+		CreatedAt: r.CreatedAt,
+		UpdatedAt: r.UpdatedAt,
+
+		Port: int(r.Port),
 	}
-	route := &TCPRoute{Route: &rCopy}
-	route.Route.Config = nil
-	json.Unmarshal(*r.Config, route)
-	return route
 }
 
+func (r Route) routeForDB() interface{} {
+	switch r.Type {
+	case "http":
+		return r.HTTPRoute()
+	case "tcp":
+		return r.TCPRoute()
+	default:
+		panic(fmt.Sprintf("unknown table name: %q", r.Type))
+	}
+}
+
+// HTTPRoute is an HTTP Route.
 type HTTPRoute struct {
-	*Route  `json:"-"`
+	// TODO(bgentry): remove json tags here, or make this serialize properly with
+	// a Type field.
```

Comment:
Basically, in order for either `HTTPRoute` or `TCPRoute` to serialize properly, they need to also have a `type` field added to the fields listed below. There are 2 ways I can think of to do that:

* Add a `MarshalJSON()` method that uses a private (anonymous?) struct specifically for serialization, which also includes a `type` field
* Add a `MarshalJSON()` method that calls `ToRoute()` and then marshals the result of that

Otherwise we can just eliminate these tags altogether, which means you always need to serialize directly with a `Route`.